esp/services/ws_codesign/ws_codesignService.cpp

```diff
@@ -0,0 +1,92 @@
+
+    HPCC SYSTEMS software Copyright (C) 2019 HPCC Systems®.
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+############################################################################## */
+
+#include "ws_codesignService.hpp"
+#include "jutil.hpp"
+
+Cws_codesignEx::Cws_codesignEx()
+{
+}
+
+Cws_codesignEx::~Cws_codesignEx()
+{
+}
+
+void Cws_codesignEx::init(IPropertyTree *cfg, const char *process, const char *service)
+{
+    if(cfg == nullptr)
+        throw MakeStringException(-1, "Cannot initialize Cws_codesignEx, cfg is NULL");
+
+    StringBuffer xpath;
+    xpath.appendf("Software/EspProcess[@name=\"%s\"]/EspService[@name=\"%s\"]", process, service);
+    m_serviceCfg.setown(cfg->getPropTree(xpath.str()));
+}
+
+bool Cws_codesignEx::onSign(IEspContext &context, IEspSignRequest &req, IEspSignResponse &resp)
+{
+    resp.setRetCode(-1);
+
+    const char* keyid = req.getKeyID();
+    const char* text = req.getText();
+    if (!keyid || !*keyid || !text || !*text)
+    {
+        resp.setErrMsg("Please provide both KeyID and Text");
+        return false;
+    }
+    int keylen = strlen(keyid);
+    if(keylen != 16 && keylen != 40)
+    {
+        resp.setErrMsg("Invalid KeyID");
+        return false;
+    }
+
+    StringBuffer cmd, output, errmsg;
+
+    cmd.appendf("gpg --list-keys %s", keyid);
+    int ret = runExternalCommand(output, errmsg, cmd.str(), "");
+    if (ret != 0 || strstr(output.str(), keyid) == nullptr)
+    {
+        resp.setErrMsg("Key not found");
+        return false;
+    }
+
+    output.clear();
+    errmsg.clear();
+    runExternalCommand(output, errmsg, "gpg-connect-agent", "RELOADAGENT");
+
+    output.clear();
+    errmsg.clear();
+    cmd.clear().appendf("gpg --clearsign --default-key %s --pinentry-mode loopback --yes --batch", keyid);
+    StringBuffer passbuf(req.getKeyPass()), input;
+    if (passbuf.trim().length() > 0)
+    {
+        input.appendf("%s\n", passbuf.str());
+        cmd.append(" --passphrase-fd 0 --pinentry-mode loopback");
+    }
+    input.append(text);
+    ret = runExternalCommand(output, errmsg, cmd.str(), input.str());
+    if (ret != 0 || output.length() == 0)
+    {
+        UERRLOG("gpg clearsign error: [%d] %s\nOutput: n%s", ret, errmsg.str(), output.str());
+        resp.setErrMsg("Failed to sign text, please check service log for details");
+        return false;
+    }
+
+    resp.setRetCode(0);
+    resp.setSignedText(output.str());
```

Comment:
Does ESP automatically encode the response so it's valid in json or xml?

Reply:
Yes it does. But as a result, you need to xml-decode the result before using it. I added xslt transform to the response page GUI, from which the signed text can be copy-pasted directly from the web interface.